### ISSUE TITLE
 Integration Tests: Fail in Docker Container : Solution root could not be located using application root 

### ISSUE THREAD

**Issue by dcs3spp** [dotnet/aspnetcore#27627]
I am learning ASP.NET Core 3.1 and have managed to create my own [WebApplicationFactory](https://docs.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.mvc.testing.webapplicationfactory-1?view=aspnetcore-3.0) class with the [ConfigureWebHost](https://docs.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.mvc.testing.webapplicationfactory-1.configurewebhost?view=aspnetcore-3.0) method overriden as follows:

```C#
/// <summary>
/// Set the content root to path relative to sln, e.g. Src/WebApp
/// Configure the webhost using appsettings.json from the path where the assembly is located for 
/// the startup class. The Autofac container is configured here also.
/// </summary>
/// <seealso cref="Microsoft.AspNetCore.Mvc.Testing.ConfigureWebHost(IWebHostBuilder)"/></seealso>
protected override void ConfigureWebHost(IWebHostBuilder builder)
{
    string path = Assembly.GetAssembly(typeof(WebApiTestFactory<TStartupClass>)).Location;

    builder.UseSolutionRelativeContentRoot(_contentRoot)
        .ConfigureAppConfiguration(cb =>
        {
            cb.AddJsonFile($"{Path.GetDirectoryName(path)}/appsettings.json", optional: false)
                .AddEnvironmentVariables();
        })
        .ConfigureServices(services => services.AddAutofac());

    base.ConfigureWebHost(builder);
}
```

This uses ```UseSolutionRelativeContentRoot``` to set the path for content root that is relative to the directory containing *.sln files.

My integration test runs successfully within local development environment with a startup class inherited from the SUT startup.

However, when I run the test within a docker container an ```InvalidOperationException``` is thrown: ```Solution root could not be located using application root```

I also have an xunit.runner.json to disable shadow copy:
```
{
  "shadowCopy": false
}
```

Any ideas how I can solve this?

**Exception**
```bash
A total of 1 test files matched the specified pattern.
/Tests/FunctionalTests/WebApp.FunctionalTests/bin/Debug/netcoreapp3.1/WebApp.FunctionalTests.dll
[xUnit.net 00:00:00.00] xUnit.net VSTest Adapter v2.4.1 (64-bit .NET Core 3.1.9)
[xUnit.net 00:00:00.88]   Discovering: WebApp.FunctionalTests
[xUnit.net 00:00:00.94]   Discovered:  WebApp.FunctionalTests
[xUnit.net 00:00:00.95]   Starting:    WebApp.FunctionalTests
[xUnit.net 00:00:01.17]     WebApp.FunctionalTests.ApiTest.WebApp_ApiController_DownloadImage [FAIL]
[xUnit.net 00:00:01.17]       System.InvalidOperationException : Solution root could not be located using application root /Tests/FunctionalTests/WebApp.FunctionalTests/bin/Debug/netcoreapp3.1/.
[xUnit.net 00:00:01.18]       Stack Trace:
[xUnit.net 00:00:01.18]            at Microsoft.AspNetCore.TestHost.WebHostBuilderExtensions.UseSolutionRelativeContentRoot(IWebHostBuilder builder, String solutionRelativePath, String applicationBasePath, String solutionName)
[xUnit.net 00:00:01.18]            at Microsoft.AspNetCore.Mvc.Testing.WebApplicationFactory`1.SetContentRoot(IWebHostBuilder builder)
[xUnit.net 00:00:01.18]            at Microsoft.AspNetCore.Mvc.Testing.WebApplicationFactory`1.EnsureServer()
[xUnit.net 00:00:01.18]            at Microsoft.AspNetCore.Mvc.Testing.WebApplicationFactory`1.CreateDefaultClient(DelegatingHandler[] handlers)
[xUnit.net 00:00:01.18]            at Microsoft.AspNetCore.Mvc.Testing.WebApplicationFactory`1.CreateDefaultClient(Uri baseAddress, DelegatingHandler[] handlers)
[xUnit.net 00:00:01.18]            at Microsoft.AspNetCore.Mvc.Testing.WebApplicationFactory`1.CreateClient(WebApplicationFactoryClientOptions options)
[xUnit.net 00:00:01.18]            at Microsoft.AspNetCore.Mvc.Testing.WebApplicationFactory`1.CreateClient()
[xUnit.net 00:00:01.18]         /Tests/FunctionalTests/WebApp.FunctionalTests/ApiTest.cs(45,0): at WebApp.FunctionalTests.ApiTest.WebApp_ApiController_DownloadImage()
[xUnit.net 00:00:01.18]         --- End of stack trace from previous location where exception was thrown ---
[xUnit.net 00:00:01.19]   Finished:    WebApp.FunctionalTests
  X WebApp.FunctionalTests.ApiTest.WebApp_ApiController_DownloadImage [93ms]
  Error Message:
   System.InvalidOperationException : Solution root could not be located using application root /Tests/FunctionalTests/WebApp.FunctionalTests/bin/Debug/netcoreapp3.1/.
  Stack Trace:
     at Microsoft.AspNetCore.TestHost.WebHostBuilderExtensions.UseSolutionRelativeContentRoot(IWebHostBuilder builder, String solutionRelativePath, String applicationBasePath, String solutionName)
   at Microsoft.AspNetCore.Mvc.Testing.WebApplicationFactory`1.SetContentRoot(IWebHostBuilder builder)
   at Microsoft.AspNetCore.Mvc.Testing.WebApplicationFactory`1.EnsureServer()
   at Microsoft.AspNetCore.Mvc.Testing.WebApplicationFactory`1.CreateDefaultClient(DelegatingHandler[] handlers)
   at Microsoft.AspNetCore.Mvc.Testing.WebApplicationFactory`1.CreateDefaultClient(Uri baseAddress, DelegatingHandler[] handlers)
   at Microsoft.AspNetCore.Mvc.Testing.WebApplicationFactory`1.CreateClient(WebApplicationFactoryClientOptions options)
   at Microsoft.AspNetCore.Mvc.Testing.WebApplicationFactory`1.CreateClient()
   at WebApp.FunctionalTests.ApiTest.WebApp_ApiController_DownloadImage() in /Tests/FunctionalTests/WebApp.FunctionalTests/ApiTest.cs:line 45
--- End of stack trace from previous location where exception was thrown ---

Test Run Failed.
Total tests: 1
     Failed: 1
 Total time: 2.0426 Seconds
/usr/share/dotnet/sdk/3.1.403/Microsoft.TestPlatform.targets(32,5): error MSB4181: The "Microsoft.TestPlatform.Build.Tasks.VSTestTask" task returned false but did not log an error. [/Tests/FunctionalTests/WebApp.FunctionalTests/WebApp.FunctionalTests.csproj]
```

**Test Project**
```xml
<Project Sdk="Microsoft.NET.Sdk.Web">

  <PropertyGroup>
    <TargetFramework>netcoreapp3.1</TargetFramework>

    <IsPackable>false</IsPackable>
  </PropertyGroup>

  <ItemGroup>
    <Content Update="xunit.runner.json">
      <CopyToOutputDirectory>Always</CopyToOutputDirectory>
    </Content>
  </ItemGroup>

  <ItemGroup>
    <PackageReference Include="Autofac" Version="6.0.0" />
    <PackageReference Include="Microsoft.AspNetCore" Version="2.2.0" />
    <PackageReference Include="Microsoft.NET.Test.Sdk" Version="16.2.0" />
    <PackageReference Include="xunit" Version="2.4.1" />
    <PackageReference Include="xunit.runner.visualstudio" Version="2.4.1" />
    <PackageReference Include="coverlet.collector" Version="1.0.1" />
  </ItemGroup>

  <ItemGroup>
    <ProjectReference Include="..\..\..\Src\WebApp\WebApp.csproj" />
    <ProjectReference Include="..\..\Functional\Utilities\WebApp.Functional.Utilities.csproj" />
    <ProjectReference Include="..\..\..\Src\WebApp.S3.Contracts\WebApp.S3.Contracts.csproj" />
  </ItemGroup>

</Project>
```

Also added Microsoft.AspNetCore.Mvc.Testing dependency (as detailed [here](https://docs.microsoft.com/en-us/aspnet/core/test/integration-tests?view=aspnetcore-3.1#test-app-prerequisites)) to the test project and still get the same result:
```xml
 <PackageReference Include="Microsoft.AspNetCore.Mvc.Testing" Version="3.1.9" />
```

_Originally posted by @dcs3spp in https://github.com/dotnet/aspnetcore/discussions/27613_

Also tried adding WebApplicationFactoryContentRoot attribute in Test Project:
```C#
using System;
using System.IO;
using System.Reflection;
using System.Text;
using System.Threading.Tasks;

using Microsoft.AspNetCore.Hosting;
using Autofac.Extensions.DependencyInjection;
using Microsoft.Extensions.DependencyInjection;
using Microsoft.AspNetCore.Mvc.Testing;
using Xunit;
using Xunit.Abstractions;

using WebApp.S3.Contracts;
using WebApp.Functional.Utilities;

/// Key should match FullName assembly attributed of TStartup : WebApplicationFactory<TStartup> 
/// https://docs.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.mvc.testing.webapplicationfactorycontentrootattribute?view=aspnetcore-3.0
[assembly: WebApplicationFactoryContentRoot(
    key: "WebApp.FunctionalTests, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null",
    contentRootPath: "../../../../../../Src/WebApp",
    contentRootTest: "Program.cs",
    priority: "1")]

namespace WebApp.FunctionalTests
{
    public class ApiTest
    {
        private TestingBase<WebAppTestStartup> _testing;
        private string _settingsFile;
        
        
        private string GetSettingsFile() 
        {
            const string AspNetCoreEnvironment="ASPNETCORE_ENVIRONMENT";
            const string DefaultSettingsFile = "appsettings.Local.json";
            
            var envVar = System.Environment.GetEnvironmentVariable(AspNetCoreEnvironment);

            if(envVar != null)
                return $"appsettings.{envVar}.json";
            else
                return DefaultSettingsFile;
        }

        /// <summary>
        /// Create Web Application Factory with content root at Src/WebApp using
        /// appsettings file derived from ASPNETCORE_ENVIRONMENT variable. If
        /// the variable is unset then defaults to appsettings.Local.json
        /// </summary>
        /// <param name="output">xUnit output stream</param>
        public ApiTest(ITestOutputHelper output)
        {
            const string contentRoot = "Src/WebApp";
            
            _settingsFile = GetSettingsFile();
            
            string startupPath = Assembly.GetAssembly(typeof(WebApiTestFactory<WebAppTestStartup>)).Location;
            string settingsFilePath = $"{Path.GetDirectoryName(startupPath)}/{_settingsFile}";

            _testing = new TestingBase<WebAppTestStartup>(output, contentRoot, settingsFilePath);
        }

        [Fact]
        public async Task WebApp_ApiController_DownloadImage()
        {
            const string TestData = "test";
            const string ObjectName = "objectName";
            byte[] Payload = Encoding.UTF8.GetBytes(TestData);

            // upload a sample image
            using (var client = _testing.Factory.CreateClient())
            using (var scope = _testing.Factory.Server.Host.Services.CreateScope())
            using (var byteStream = new MemoryStream(Payload))
            using (var stream = new BufferedStream(byteStream))
            {
                var s3Client = scope.ServiceProvider.GetRequiredService<IS3Service>();
                await s3Client.UploadAsync(stream, ObjectName);

                var response = await client.GetAsync(ApiRoutes.Get.Image(ObjectName));
                response.EnsureSuccessStatusCode();

                string result = await response.Content.ReadAsStringAsync();

                Assert.Equal(TestData, result);
            }
        }
    }
}
```

**Comment by javiercn** [comment]
@dcs3spp thanks for contacting us.

Can you provide the docker file you are using to run the tests? It is likely that the paths that we map automatically are not valid in this context depending on how you are running the tests within the container.

**Comment by dcs3spp** [comment]
@javiercn thanks for responding .... docker file included below:

**Docker**
```Docker
##
# Dockerfile for building, testing and publishing Release
# Intended for use in conjunction with docker buildkit
#
# Stages:
# base:             Contains alpine packages and dotnet tools
# build-prod:       Minimal depencencies restored
# build-test:       Development dependencies restored
# test:             If targeted, enables optional entrypoint for testing
# publish:          Publish release for WebApp/WebApp.csproj with semantic version number
#
# Usage:
# Test:             DOCKER_BUILDKIT=1 docker build --target test -f Docker/Test/Dockerfile -t webapp .
# Release:          DOCKER_BUILDKIT=1 docker build -f Docker/Test/Dockerfile -t webapp .
#
# TODO:
# 1. RUN for published image needs to find settings from env vars in production
# 2. Functional Test error "Solution root could not be located using application root"
##


##
# Base Image
# Image: SDK
# 
# - Installs reportgenerator global tool and updates path
# - Sets PS1 prompt coloured to include user@host and working directory
##
FROM mcr.microsoft.com/dotnet/core/sdk:3.1-alpine AS base

ENV PATH="$PATH:/root/.dotnet/tools"
ENV PS1A="[\d \[\e[33m\]\t\[\e[m\]] \[\e[31m\]\u\[\e[m\]@\[\e[31m\]\h\[\e[m\] \[\e[36m\]\w\[\e[m\]\n $ "

RUN apk add --update --no-cache git bash curl zip \
    && echo 'PS1=$PS1A' >> ~/.bashrc \
    && dotnet tool install --global dotnet-reportgenerator-globaltool --version 4.6.6


##
# Perform restore for production dependencies
##
FROM base AS build-prod

COPY "WebApp.Production.sln" "WebApp.Production.sln"

WORKDIR /Src

COPY "Src/WebApp/WebApp.csproj" "WebApp/WebApp.csproj"
COPY "Src/WebApp.BackgroundServices/WebApp.BackgroundServices.csproj" "WebApp.BackgroundServices/WebApp.BackgroundServices.csproj"
COPY "Src/WebApp.Data/WebApp.Data.csproj" "WebApp.Data/WebApp.Data.csproj"
COPY "Src/WebApp.Data.Serializers/WebApp.Data.Serializers.csproj" "WebApp.Data.Serializers/WebApp.Data.Serializers.csproj"
COPY "Src/WebApp.Data.Serializers.AutofacModule/WebApp.Data.Serializers.AutofacModule.csproj" "WebApp.Data.Serializers.AutofacModule/WebApp.Data.Serializers.AutofacModule.csproj"
COPY "Src/WebApp.Data.Serializers.Contracts/WebApp.Data.Serializers.Contracts.csproj" "WebApp.Data.Serializers.Contracts/WebApp.Data.Serializers.Contracts.csproj"
COPY "Src/WebApp.Extensions/WebApp.Extensions.csproj" "WebApp.Extensions/WebApp.Extensions.csproj"
COPY "Src/WebApp.Kafka/WebApp.Kafka.csproj" "WebApp.Kafka/WebApp.Kafka.csproj"
COPY "Src/WebApp.Kafka.AutofacModule/WebApp.Kafka.AutofacModule.csproj" "WebApp.Kafka.AutofacModule/WebApp.Kafka.AutofacModule.csproj"
COPY "Src/WebApp.Mqtt/WebApp.Mqtt.csproj" "WebApp.Mqtt/WebApp.Mqtt.csproj"
COPY "Src/WebApp.Mqtt.AutofacModule/WebApp.Mqtt.AutofacModule.csproj" "WebApp.Mqtt.AutofacModule/WebApp.Mqtt.AutofacModule.csproj"
COPY "Src/WebApp.Realtime.SignalR/WebApp.Realtime.SignalR.csproj" "WebApp.Realtime.SignalR/WebApp.Realtime.SignalR.csproj"
COPY "Src/WebApp.Repository/WebApp.Repository.csproj" "WebApp.Repository/WebApp.Repository.csproj"
COPY "Src/WebApp.Repository.AutofacModule/WebApp.Repository.AutofacModule.csproj" "WebApp.Repository.AutofacModule/WebApp.Repository.AutofacModule.csproj"
COPY "Src/WebApp.Repository.Contracts/WebApp.Repository.Contracts.csproj" "WebApp.Repository.Contracts/WebApp.Repository.Contracts.csproj"
COPY "Src/WebApp.S3/WebApp.S3.csproj" "WebApp.S3/WebApp.S3.csproj"
COPY "Src/WebApp.S3.AutofacModule/WebApp.S3.AutofacModule.csproj" "WebApp.S3.AutofacModule/WebApp.S3.AutofacModule.csproj"
COPY "Src/WebApp.S3.Contracts/WebApp.S3.Contracts.csproj" "WebApp.S3.Contracts/WebApp.S3.Contracts.csproj"

WORKDIR /
RUN dotnet restore -r linux-musl-x64 --no-cache "WebApp.Production.sln"

WORKDIR /Src
COPY Src/ .


##
# Build Dependencies for testing
##
FROM base AS build-test

COPY "WebApp.sln" "WebApp.sln"

WORKDIR /Src
COPY "Src/WebApp/WebApp.csproj" "WebApp/WebApp.csproj"
COPY "Src/WebApp.BackgroundServices/WebApp.BackgroundServices.csproj" "WebApp.BackgroundServices/WebApp.BackgroundServices.csproj"
COPY "Src/WebApp.Data/WebApp.Data.csproj" "WebApp.Data/WebApp.Data.csproj"
COPY "Src/WebApp.Data.Serializers/WebApp.Data.Serializers.csproj" "WebApp.Data.Serializers/WebApp.Data.Serializers.csproj"
COPY "Src/WebApp.Data.Serializers.AutofacModule/WebApp.Data.Serializers.AutofacModule.csproj" "WebApp.Data.Serializers.AutofacModule/WebApp.Data.Serializers.AutofacModule.csproj"
COPY "Src/WebApp.Data.Serializers.Contracts/WebApp.Data.Serializers.Contracts.csproj" "WebApp.Data.Serializers.Contracts/WebApp.Data.Serializers.Contracts.csproj"
COPY "Src/WebApp.Extensions/WebApp.Extensions.csproj" "WebApp.Extensions/WebApp.Extensions.csproj"
COPY "Src/WebApp.Kafka/WebApp.Kafka.csproj" "WebApp.Kafka/WebApp.Kafka.csproj"
COPY "Src/WebApp.Kafka.AutofacModule/WebApp.Kafka.AutofacModule.csproj" "WebApp.Kafka.AutofacModule/WebApp.Kafka.AutofacModule.csproj"
COPY "Src/WebApp.Mqtt/WebApp.Mqtt.csproj" "WebApp.Mqtt/WebApp.Mqtt.csproj"
COPY "Src/WebApp.Mqtt.AutofacModule/WebApp.Mqtt.AutofacModule.csproj" "WebApp.Mqtt.AutofacModule/WebApp.Mqtt.AutofacModule.csproj"
COPY "Src/WebApp.Realtime.SignalR/WebApp.Realtime.SignalR.csproj" "WebApp.Realtime.SignalR/WebApp.Realtime.SignalR.csproj"
COPY "Src/WebApp.Repository/WebApp.Repository.csproj" "WebApp.Repository/WebApp.Repository.csproj"
COPY "Src/WebApp.Repository.AutofacModule/WebApp.Repository.AutofacModule.csproj" "WebApp.Repository.AutofacModule/WebApp.Repository.AutofacModule.csproj"
COPY "Src/WebApp.Repository.Contracts/WebApp.Repository.Contracts.csproj" "WebApp.Repository.Contracts/WebApp.Repository.Contracts.csproj"
COPY "Src/WebApp.S3/WebApp.S3.csproj" "WebApp.S3/WebApp.S3.csproj"
COPY "Src/WebApp.S3.AutofacModule/WebApp.S3.AutofacModule.csproj" "WebApp.S3.AutofacModule/WebApp.S3.AutofacModule.csproj"
COPY "Src/WebApp.S3.Contracts/WebApp.S3.Contracts.csproj" "WebApp.S3.Contracts/WebApp.S3.Contracts.csproj"

WORKDIR /Tests
COPY "Tests/Fixtures/WebApp.Testing.Fixtures/WebApp.Testing.Fixtures.csproj"  "Fixtures/WebApp.Testing.Fixtures/WebApp.Testing.Fixtures.csproj"
COPY "Tests/Functional/Utilities/WebApp.Functional.Utilities.csproj" "Functional/Utilities/WebApp.Functional.Utilities.csproj"
COPY "Tests/IntegrationTests/WebApp.Data.Serializer.IntegrationTests/WebApp.Data.Serializer.IntegrationTests.csproj" "IntegrationTests/WebApp.Data.Serializer.IntegrationTests/WebApp.Data.Serializer.IntegrationTests.csproj"
COPY "Tests/IntegrationTests/WebApp.Data.Serializers.AutofacModule.IntegrationTests/WebApp.Data.Serializers.AutofacModule.IntegrationTests.csproj" "IntegrationTests/WebApp.Data.Serializers.AutofacModule.IntegrationTests/WebApp.Data.Serializers.AutofacModule.IntegrationTests.csproj"
COPY "Tests/IntegrationTests/WebApp.IntegrationTests/WebApp.IntegrationTests.csproj" "IntegrationTests/WebApp.IntegrationTests/WebApp.IntegrationTests.csproj"
COPY "Tests/IntegrationTests/WebApp.Kafka.AutofacModule.IntegrationTests/WebApp.Kafka.AutofacModule.IntegrationTests.csproj" "IntegrationTests/WebApp.Kafka.AutofacModule.IntegrationTests/WebApp.Kafka.AutofacModule.IntegrationTests.csproj"
COPY "Tests/IntegrationTests/WebApp.Mqtt.IntegrationTests/WebApp.Mqtt.IntegrationTests.csproj" "IntegrationTests/WebApp.Mqtt.IntegrationTests/WebApp.Mqtt.IntegrationTests.csproj"
COPY "Tests/IntegrationTests/WebApp.Mqtt.AutofacModule.IntegrationTests/WebApp.Mqtt.AutofacModule.IntegrationTests.csproj" "IntegrationTests/WebApp.Mqtt.AutofacModule.IntegrationTests/WebApp.Mqtt.AutofacModule.IntegrationTests.csproj"
COPY "Tests/IntegrationTests/WebApp.Repository.AutofacModule.IntegrationTests/WebApp.Repository.AutofacModule.IntegrationTests.csproj" "IntegrationTests/WebApp.Repository.AutofacModule.IntegrationTests/WebApp.Repository.AutofacModule.IntegrationTests.csproj"
COPY "Tests/IntegrationTests/WebApp.S3.AutofacModule.IntegrationTests/WebApp.S3.AutofacModule.IntegrationTests.csproj" "IntegrationTests/WebApp.S3.AutofacModule.IntegrationTests/WebApp.S3.AutofacModule.IntegrationTests.csproj"
COPY "Tests/Mocks/WebApp.Testing.Mocks/WebApp.Testing.Mocks.csproj" "Mocks/WebApp.Testing.Mocks/WebApp.Testing.Mocks.csproj"
COPY "Tests/UnitTests/WebApp.BackgroundServices.UnitTests/WebApp.BackgroundServices.UnitTests.csproj" "UnitTests/WebApp.BackgroundServices.UnitTests/WebApp.BackgroundServices.UnitTests.csproj" 
COPY "Tests/UnitTests/WebApp.Data.Serializers.UnitTests/WebApp.Data.Serializers.UnitTests.csproj"  "UnitTests/WebApp.Data.Serializers.UnitTests/WebApp.Data.Serializers.UnitTests.csproj"
COPY "Tests/UnitTests/WebApp.Data.UnitTests/WebApp.Data.UnitTests.csproj" "UnitTests/WebApp.Data.UnitTests/WebApp.Data.UnitTests.csproj"
COPY "Tests/UnitTests/WebApp.Extensions.UnitTests/WebApp.Extensions.UnitTests.csproj" "UnitTests/WebApp.Extensions.UnitTests/WebApp.Extensions.UnitTests.csproj"
COPY "Tests/UnitTests/WebApp.Kafka.UnitTests/WebApp.Kafka.UnitTests.csproj" "UnitTests/WebApp.Kafka.UnitTests/WebApp.Kafka.UnitTests.csproj"
COPY "Tests/UnitTests/WebApp.Mqtt.UnitTests/WebApp.Mqtt.UnitTests.csproj" "UnitTests/WebApp.Mqtt.UnitTests/WebApp.Mqtt.UnitTests.csproj"
COPY "Tests/UnitTests/WebApp.Realtime.SignalR.UnitTests/WebApp.Realtime.SignalR.UnitTests.csproj" "UnitTests/WebApp.Realtime.SignalR.UnitTests/WebApp.Realtime.SignalR.UnitTests.csproj"
COPY "Tests/UnitTests/WebApp.Repository.UnitTests/WebApp.Repository.UnitTests.csproj" "UnitTests/WebApp.Repository.UnitTests/WebApp.Repository.UnitTests.csproj"
COPY "Tests/UnitTests/WebApp.S3.UnitTests/WebApp.S3.UnitTests.csproj" "UnitTests/WebApp.S3.UnitTests/WebApp.S3.UnitTests.csproj"
COPY "Tests/UnitTests/WebApp.UnitTests/WebApp.UnitTests.csproj" "UnitTests/WebApp.UnitTests/WebApp.UnitTests.csproj"
COPY "Tests/FunctionalTests/WebApp.FunctionalTests/WebApp.FunctionalTests.csproj" "FunctionalTests/WebApp.FunctionalTests/WebApp.FunctionalTests.csproj"

WORKDIR /
RUN dotnet restore -r linux-musl-x64 --no-cache "WebApp.sln"

COPY . .


##
# Optional entrypoint for testing
# Build with `docker build --target test`
##
FROM build-test AS test
ENTRYPOINT ["dotnet", "test", "--logger:trx", "WebApp.sln"]


##
# Publish a Release of WebApp/WebApp.csproj to /app
# Build arguments: 
# VERSION:   Semnatic version of release, defaults to 0.0.0
##
FROM build-prod as publish
ARG VERSION="0.0.0"
RUN dotnet publish "WebApp/WebApp.csproj" \
    -r linux-musl-x64 \
    --output /app \
    --configuration Release \
    --self-contained true \
    --no-restore \
    /p:Version=$VERSION


##
# Copy published web app and start
##
FROM mcr.microsoft.com/dotnet/core/aspnet:3.1-alpine AS final
WORKDIR /app
COPY --from=publish /app .
ENTRYPOINT ["dotnet", "WebApp.dll"]

```

**Comment by javiercn** [comment]
@dcs3spp It's a bit hard to follow based on the docker file, so I'll try to give you pointers instead to solve it. You need to make sure that there is a `WebApplicationFactoryContentRoot` attribute that is able to find the `${ContentRoot}/{File}` that you specify on the attribute. The steps to do so are:
* Determine where your test assembly lives in the container filesystem.
* Determine where the target project lives within the container filesystem.
* Add a relative path from where the test assembly is to where the application needs to be.
  * If you are not sure about this, build the container, run a command-line to the test step container and look at the filesystem to be sure.
  * Beware of casing differences.
* Create the attribute pointing to the content root folder with either an absolute path or a path relative from where the test assembly lives in the container.
* Use a file to make sure the infrastructure can detect when that attribute applies.

**Comment by dcs3spp** [comment]
@javiercn Thanks, appreciated :) 
Docker container file structure looks like it is matching local solution file structure, but will double check.

Yes, I read the [documentation](https://docs.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.mvc.testing.webapplicationfactorycontentrootattribute?view=aspnetcore-3.0) for ```WebApplicationFactoryContentRoot``` as listed in test project code above.

I am not sure if I am using it correctly, cannot find an example with an existing integration test. So I recently created a smaller test and opened another [discussion](https://docs.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.mvc.testing.webapplicationfactorycontentrootattribute?view=aspnetcore-3.0) before dabbling with the test in docker.

**Comment by dcs3spp** [comment]
@javiercn Hi, I have since created a smaller repository that highlights this issue with a much simpler Dockerfile. As far as I can tell the docker image file structure mirrors the structure of the local development environment.

```bash
git clone https://github.com/dcs3spp/WebApplicationFactoryContentRoot.git

cd WebApplicationFactoryContentRoot/Tests/FunctionalTests/WebApp.FunctionalTests.UseSolutionRelativeContentRoot

# run test locally succeeds
dotnet test

# move to root folder to to build and run docker container containing this test
cd ../../../

# build docker image called docker-webapp
DOCKER_BUILDKIT=1 docker build -f Docker/Dockerfile -t docker-webapp .

# run a docker container containing this test - fails given solution root error
docker run --rm --entrypoint=bash -it docker-webapp
```

**Comment by joaocpribeiro** [comment]
@javiercn I am still experiencing the referenced error message ("Solution root could not be located using application root").
I verified that the path and the marker are correctly set.
My attribute is set as the following:
```
[assembly: WebApplicationFactoryContentRoot
    ("<my API assembly name>, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null",
    "./ArtifactDirectoryName", "appsettings.json", "1")]
```
I did some debugging and noticed that the version (in the assembly full name) is incorrect. My Dockerfile is putting the API's publish output in the ArtifactDirectoryName, and the version of the assembly is changing each time docker is built.
Do you have any suggestion on how to fix or workaround this issue?

In my mind it would make sense to expect the assembly name as the "key" of WebApplicationFactoryContentRootAttrbute instead of the full name, but I guess there are good reasons for why it is as it is.

**Comment by joaocpribeiro** [comment]
I found a workaround. I removed the WebApplicationFactoryContentRootAttribute and created a fake (and empty) .sln file in my docker container. I also had to create a directory with the name of the API's project. The rest is magic. 😄 
Anyway, if there is a better approach, I would be happy to read about it.
